rkyv/src/niche/niched_option.rs

```diff
@@ -0,0 +1,285 @@
+//! A niched archived `Option<T>` that uses less space based on a niching
+//! [`Decider`].
+
+use core::{cmp, fmt, mem::ManuallyDrop, ops::Deref};
+
+use munge::munge;
+use rancor::Fallible;
+
+use super::decider::Decider;
+use crate::{seal::Seal, Archive, Archived, Place, Portable, Serialize};
+
+/// A niched archived `Option<T>`.
+///
+/// It uses less space by storing the `None` variant in a custom way based on
+/// `D`.
+#[derive(Portable)]
+#[rkyv(crate)]
+#[cfg_attr(feature = "bytecheck", derive(bytecheck::CheckBytes))]
+#[repr(transparent)]
+pub struct NichedOption<T, D>
+where
+    T: Archive,
+    D: Decider<T> + ?Sized,
+{
+    repr: Repr<T, D>,
+}
+
+#[repr(C)]
+#[derive(Portable)]
+#[rkyv(crate)]
+union Repr<T, D>
+where
+    T: Archive,
+    D: Decider<T> + ?Sized,
+{
+    some: ManuallyDrop<<T as Archive>::Archived>,
+    niche: ManuallyDrop<<D as Decider<T>>::Niched>,
+}
+
+impl<T, D> Repr<T, D>
+where
+    T: Archive,
+    D: Decider<T> + ?Sized,
+{
+    /// Compile-time check to make sure that the niched type is not greater than
+    /// the archived type.
+    ///
+    /// ```compile_fail
+    /// use rkyv::{
+    ///     niche::{decider::Decider, niched_option::NichedOption},
+    ///     Archived, Place,
+    /// };
+    ///
+    /// type T = u16;
+    /// type N = u32;
+    ///
+    /// struct UselessDecider;
+    ///
+    /// unsafe impl Decider<T> for UselessDecider {
+    ///     type Niched = Archived<N>;
+    ///
+    ///     fn is_niched(_: &Archived<N>) -> bool {
+    ///         false
+    ///     }
+    ///
+    ///     fn resolve_niche(_: Place<Self::Niched>) {}
+    /// }
+    ///
+    /// let archived: Archived<N> = 456.into();
+    /// let niched: &NichedOption<T, UselessDecider> =
+    ///     unsafe { std::mem::transmute(&archived) };
+    /// let _ = niched.is_none(); // <- size check = compile error
+    /// ```
+    const NICHE_SIZE_CHECK: () = {
+        if size_of::<<D as Decider<T>>::Niched>()
+            > size_of::<<T as Archive>::Archived>()
+        {
+            panic!(
+                "`D::Niched` is greater than `T::Archived` and thus useless \
+                 for niching"
+            );
+        }
+    };
+}
+
+#[cfg(feature = "bytecheck")]
+const _: () = {
+    use crate::bytecheck::CheckBytes;
+
+    unsafe impl<T, D, C> CheckBytes<C> for Repr<T, D>
+    where
+        T: Archive<Archived: CheckBytes<C>>,
+        D: Decider<T, Niched: CheckBytes<C>> + ?Sized,
+        C: Fallible + ?Sized,
+    {
+        unsafe fn check_bytes(
+            value: *const Self,
+            context: &mut C,
+        ) -> Result<(), C::Error> {
+            unsafe { <D::Niched>::check_bytes(&*(*value).niche, context)? };
```

Comment:
There's a very minor risk that checking the niched type and then the un-niched type could cause problems if they both mutate the context. Do you think it would be useful to require `D::Niched: CheckBytes<()>`? (meant to indicate "this type can be checked without any additional context").

I could see someone wanting to access the context for benign reasons though (like "runtime niching" or something), so maybe it's just best to document that this type may check the niched type as well as the un-niched type.

Reply:
True, good catch.

I think it might just be sufficient to check the niched type and be done with it.
- If it fails we return an error
- If it succeeds and `D::is_niched` would return true then we return ok.
- If it succeeds and `D::is_niched` would return false then the safety invariant of `Niching` as it is currently stated requires the implementor to ensure that `T` must then be safe to access.

Does that sound right?

Reply:
I agree, though it should be written down in the docs that the niched type should also `check_bytes` on any inner non-niched values, directly or indirectly. Like for `ArchivedBox<T>`, the niched repr is the same as the internal of `ArchivedBox` anyway, so double-checking is unnecessary. Same with the non-zeros and floats.